app/src/main/java/org/mozilla/fenix/HomeActivity.kt

```diff
@@ -147,7 +148,7 @@ open class HomeActivity : AppCompatActivity() {
         // Toggle off the open_link_in_private_tab pref if we are no longer set as the default browser
         // We do this on a separate thread to alleviate performance issues
         val weakReferenceContext = WeakReference(this)
-        lifecycleScope.launch {
+        lifecycleScope.launch(Dispatchers.IO) {
```

Comment:
Note to self to follow-up on which Dispatcher is appropriate here. afaik, IO is intended for blocking tasks so it may do more expensive things like create new threads to support those operations.

Reply:
There are 3 primary dispatchers:
- `Main` (The UI thread)
- `IO`
- `Default`

@colintheshots did some research and found that we should avoid `Default` as much as possible. IIRC you only get 2 threads and search uses one of them already, so using `Default` causes a lot of perf issues.

For pretty much every case we use `IO`, unless we need to show something on the UI.

Reply:
> For pretty much every case we use IO, unless we need to show something on the UI.

I'm concerned about this from a performance perspective: in my understanding of the original implementation (which has since changed and I have not had time to fully review the new one yet), the `IO` dispatcher may create additional threads if none are available. Creating threads is an expensive operation but, since this dispatcher is intended for blocking tasks, the thread creation time may be insignificant compared to the blocking task and there is a need for immediate execution (e.g. start network calls ASAP).

However, for CPU-bound tasks, the time of the task may not be worth the creation of an entire thread. Furthermore, context switching between threads adds unnecessary CPU burden so, in some sense, it's inefficient to have too many threads for active CPU work (as opposed to the sleeping threads for blocking calls) and it's kind of a feature to have the same number of threads available as the number of CPU cores.

There is, however, a distinction the coroutine dispatchers don't make: background work intended to return to the main thread and background work that completes silently. Background work that returns to the main thread should happen ASAP while background work that completes silently can often wait.

I looked through Colin's reports – [Audit](https://docs.google.com/document/d/17WNYoS-ucL6FR8bAM64NzH5CAT_v7wIYgdHN0mQL_r0/edit) & [ANR](https://docs.google.com/document/d/1THmvymMO5HpDb34v9neECU1453DsiWCyHqgrAq3LEdc/edit#) – and I notice they're focused on startup. For startup, many background operations are intended to return to the main thread so the distinction is important: we may want to consider alternative solutions. However, I'm concerned a blanket-policy for using the IO dispatcher has additional, unnecessary inefficiencies.

Outside of startup, it's unlikely we'll run into many CPU-bound tasks simultaneously so I think the default dispatcher is still acceptable in this case (though we may want to consider the distinctions of CPU-bound tasks I mention above).

@colintheshots @hawkinsw Any thoughts here?

Reply:
CC @MarcLeclair so all the FE perf folks have context on this discussion. :) It seems like an important one